### PR TITLE
CRW support

### PR DESCRIPTION
**Pull request by pedrocr**
Support for the CIFF based Canon CRW format. This adds a CIFF parser to rawspeed and implements a CRW decoder based on the dcraw code. Support was added to the following cameras:
- D30, D60, 10D and 300D DSLRs
- Pro1, G1, G2, G3, G5 and G6 compacts
- S30, S40, S45, S50, S60, S70 compacts

I used the Adobe DNG converter and dngmeta.sh from darktable to create the cameras.xml entries. Several of the white and black points seem off though. Most other actual CRW cameras should be easy to add. rawsamples.ch has several .CRW files that don't seem to actually be the same format and won't work with this code. I suspect these are raw files created by cameras that didn't have raw to begin with and just had it retrofitted with things like CHDK.

So far the parser only works in little-endian architectures. As I commented on the PR for the BE rename I think we should handle this at FileMap/ByteStream level and make the TIFF and CIFF IFD and Entry classes architecture agnostic. It would avoid a bunch of duplication.

### CODE REVIEW

**Comment by pedrocr** [comment]
Added some BlackAreas to fix the blackpoint issues. Added the possibility of setting negative BlackArea offsets so that you can say "40 pixel vertical area 40 pixels from the right border".

File: RawSpeed/CrwDecoder.cpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      0x65,0xe9,0x72,0xe6,0x71,0x91,0x93,0xa6,0xda,0x92,0x85,0x62,
+      0xf3,0xc5,0xb2,0xa4,0x84,0xba,0x64,0xa5,0xb3,0xd2,0x81,0xe5,
+      0xd3,0xaa,0xc4,0xca,0xf2,0xb1,0xe4,0xd1,0x83,0x63,0xea,0xc3,
+      0xe2,0x82,0xf1,0xa3,0xc2,0xa1,0xc1,0xe3,0xa2,0xe1,0xff,0xff  }
+  };
+  huff[0] = makeDecoder(first_tree[table]);
+  huff[1] = makeDecoder(second_tree[table]);
+}
+
+uint32 CrwDecoder::getbithuff (ByteStream &input, int nbits, ushort16 *huff)
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp]: This looks like a standard BitPumpMSB.

File: RawSpeed/CrwDecoder.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    return bitbuf = vbits = reset = 0;
+  if (nbits == 0 || vbits < 0) return 0;
+  while (!reset && vbits < nbits && (c = ((int)input.getByte())) != EOF &&
+    !(reset = 1 && c == 0xff && ((int) input.getByte()))) {
+    bitbuf = (bitbuf << 8) + (uchar8) c;
+    vbits += 8;
+  }
+  c = bitbuf << (32-vbits) >> (32-nbits);
+  if (huff) {
+    vbits -= huff[c] >> 8;
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp]: High bits  are length

File: RawSpeed/CrwDecoder.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    c = (uchar8) huff[c];
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp]: Low 8 bits are offset info huff table

File: RawSpeed/CrwDecoder.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  getbithuff(input, -1, NULL);
+
+  for (uint32 row=0; row < height; row+=8) {
+    ushort16 *dest = (ushort16*) & mRaw->getData()[row*width*2];
+    nblocks = MIN (8, height-row) * width >> 6;
+    for (block=0; block < nblocks; block++) {
+      memset (diffbuf, 0, sizeof diffbuf);
+      for (uint32 i=0; i < 64; i++ ) {
+        leaf = getbithuff(input, *huff[i > 0], huff[i > 0]+1);
+        if (leaf == 0 && i) break;
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp:230]: Error condition?

File: RawSpeed/CrwDecoder.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        if (leaf == 0xff) continue;
+        i  += leaf >> 4;
+        len = leaf & 15;
+        if (len == 0) continue;
+        diff = getbithuff(input, len, NULL);
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp]: Get bits (not huffman encoded)

File: RawSpeed/CrwDecoder.cpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  initHuffTables (dec_table, huff);
+  uint32 offset = 540 + lowbits*height*width/4;
+  ByteStream input(mFile->getData(offset), mFile->getSize() - offset);
+  getbithuff(input, -1, NULL);
+
+  for (uint32 row=0; row < height; row+=8) {
+    ushort16 *dest = (ushort16*) & mRaw->getData()[row*width*2];
+    nblocks = MIN (8, height-row) * width >> 6;
+    for (block=0; block < nblocks; block++) {
+      memset (diffbuf, 0, sizeof diffbuf);
```
Comment (klauspost) [inline RawSpeed/CrwDecoder.cpp:227]: Delta assumed to be zero

**Comment by pedrocr** [comment]
I don't really understand BitPumpMSB to be able to make the conversion to that code. I'd like to do it though as the dcraw code is quite convoluted. Any pointers on how to go about this?

**Comment by pedrocr** [comment]
I've looked into this now and understand the BitPumpMSB code. As it stands we can't use it now because this format apparently requires skipping one or two bytes in the input whenever 0xff shows up. At least the way the rest of the code is setup that is the case.

**Comment by klauspost** [comment]
That is actually what BitPumpJPEG does, so that could maybe be used. Either way, I wouldn't mind looking through the decoder myself and doing it, but I am very short on time at the moment.

**Comment by pedrocr** [comment]
It seems to be that indeed. I'll fix it then. I've already refactored the same code for SrwDecoder using BitPumpMSB so it's mostly copy paste.

**Comment by pedrocr** [comment]
Done

**Pull request merged**